flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/window/WindowTableFunctionOperator.java

```diff
@@ -0,0 +1,112 @@
+ * {@code window_end}, {@code window_time} to indicate the assigned window.
+ *
+ * <p>Note: The operator only works for row-time.
+ *
+ * <p>Note: The operator emits per record instead of at the end of window.
+ */
+public class WindowTableFunctionOperator extends TableStreamOperator<RowData>
+        implements OneInputStreamOperator<RowData, RowData> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final WindowAssigner<TimeWindow> windowAssigner;
+    private final int rowtimeIndex;
+
+    /**
+     * The shift timezone of the window, if the proctime or rowtime type is TIMESTAMP_LTZ, the shift
+     * timezone is the timezone user configured in TableConfig, other cases the timezone is UTC
+     * which means never shift when assigning windows.
+     */
+    private final ZoneId shiftTimeZone;
+
+    /** This is used for emitting elements with a given timestamp. */
+    private transient TimestampedCollector<RowData> collector;
+
+    private transient JoinedRowData outRow;
+    private transient GenericRowData windowProperties;
+    private transient LongUnaryOperator getShiftEpochMillsFunc;
+
+    public WindowTableFunctionOperator(
+            WindowAssigner<TimeWindow> windowAssigner, int rowtimeIndex, ZoneId shiftTimeZone) {
+        checkArgument(windowAssigner.isEventTime() && rowtimeIndex >= 0);
+        this.windowAssigner = windowAssigner;
+        this.rowtimeIndex = rowtimeIndex;
+        this.shiftTimeZone = shiftTimeZone;
+
+        setChainingStrategy(ChainingStrategy.ALWAYS);
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        this.collector = new TimestampedCollector<>(output);
+        collector.eraseTimestamp();
+
+        if (shiftTimeZone.getId().equals("UTC")) {
+            getShiftEpochMillsFunc = LongUnaryOperator.identity();
+        } else {
+            getShiftEpochMillsFunc = time -> TimeWindowUtil.toEpochMills(time, shiftTimeZone);
+        }
+
+        outRow = new JoinedRowData();
+        windowProperties = new GenericRowData(3);
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        RowData inputRow = element.getValue();
+        long timestamp = inputRow.getTimestamp(rowtimeIndex, 3).getMillisecond();
+        timestamp = TimeWindowUtil.toUtcTimestampMills(timestamp, shiftTimeZone);
+        Collection<TimeWindow> elementWindows = windowAssigner.assignWindows(inputRow, timestamp);
+        for (TimeWindow window : elementWindows) {
+            windowProperties.setField(0, TimestampData.fromEpochMillis(window.getStart()));
+            windowProperties.setField(1, TimestampData.fromEpochMillis(window.getEnd()));
+            windowProperties.setField(
+                    2,
+                    TimestampData.fromEpochMillis(
+                            getShiftEpochMillsFunc.applyAsLong(window.getEnd() - 1)));
```

Comment:
Why here using `LongUnaryOperator` but above `TimeWindowUtil.toUtcTimestampMills` not use?

Reply:
@JingsongLi I removed `LongUnaryOperator` because branch predication in JAVA may cover the optimization. What do you think?